lib/galaxy-social.py

```diff
@@ -0,0 +1,152 @@
+        if metadata.get("hashtags")
+        else {}
+    )
+    markdown_content = markdown.markdown(text.strip())
+    plain_content = BeautifulSoup(markdown_content, "html.parser").get_text(
+        separator="\n"
+    )
+    return plain_content, metadata
+
+
+def process_markdown_file(file_path, processed_files):
+    content, metadata = parse_markdown_file(file_path)
+    if "preview" in sys.argv:
+        try:
+            plugins["markdown"].create_post(
+                content, [], [], metadata.get("images", []), media=metadata["media"]
+            )
+            return processed_files
+        except:
+            comment_to_github(f"Failed to create preview for {file_path}.", error=True)
+    stats = {}
+    url = {}
+    for media in metadata["media"]:
+        if file_path in processed_files and media in processed_files[file_path]:
+            stats[media] = processed_files[file_path][media]
+            continue
+        mentions = metadata.get("mentions", {}).get(media, [])
+        hashtags = metadata.get("hashtags", {}).get(media, [])
+        images = metadata.get("images", [])
+        stats[media], url[media] = plugins[media].create_post(
+            content, mentions, hashtags, images
+        )
+    url_text = "\n".join([f"[{media}]({link})" for media, link in url.items() if link])
+    comment_to_github(f"Posted to:\n\n{url_text}")
+
+    processed_files[file_path] = stats
+    print(f"Processed {file_path}: {stats}")
+    return processed_files
+
+
+def main():
+    repo = os.getenv("GITHUB_REPOSITORY")
+    pr_number = os.getenv("PR_NUMBER")
+    url = f"https://api.github.com/repos/{repo}/pulls/{pr_number}/files"
+    response = requests.get(url)
+    print(response.json(), url)
+    if response.status_code == 200:
+        changed_files = response.json()
+        for file in changed_files:
+            raw_url = file["raw_url"]
+            if raw_url.endswith(".md"):
+                response = requests.get(raw_url)
+                if response.status_code == 200:
+                    with open(file["filename"], "w") as f:
+                        f.write(response.text)
+
+    processed_files = {}
+    if os.path.exists("processed_files.json"):
+        with open("processed_files.json", "r") as file:
+            processed_files = json.load(file)
+    changed_files = os.environ.get("CHANGED_FILES")
+    if changed_files:
+        for file_path in eval(changed_files.replace("\\", "")):
+            if file_path.endswith(".md"):
+                processed_files = process_markdown_file(file_path, processed_files)
+    else:
+        for root, _, files in os.walk("posts"):
+            for filename in fnmatch.filter(files, "*.md"):
+                file_path = os.path.join(root, filename)
+                processed_files = process_markdown_file(file_path, processed_files)
+    with open("processed_files.json", "w") as file:
+        json.dump(processed_files, file)
+
+
+if __name__ == "__main__":
```

Comment:
I think this needs a bit of refactoring:
Almost all the stuff that's currently living in the main() function is github-specific and should live in an extra script outside of lib/.
The file here, should be callable as a script, too, but with either a list of files or with a folder name (like implement a --files and a --folder argument), then call a main() function that does *only* the process_markdown_file() part of the work.

The github-specific script, otoh, would do all the "which files should be handled" part, then call the main() function defined here.

The goal should be to be able to run the script here entirely outside of github, i.e. in your terminal.

Reply:
Now, you can execute the code independently of GitHub by simply running `python lib/galaxy_social.py [--files FILES [FILES ...]] [--folder FOLDER] [--preview]`.
GitHub Actions will execute the `github_run.py` file, which imports `lib.galaxy_social` as a dependency.